db_setup/app/__init__.py

```diff
@@ -16,63 +17,137 @@
-            unique_resources.append(resource)
+    # Step 3: Get existing entries from DB
+    try:
+        resources_list = Resource.query.all()
+        languages_list = Language.query.all()
+        categories_list = Category.query.all()
 
+        # Convert to dict for quick lookup
+        existing_resources = {r.url:r for r in resources_list}
+        language_dict = {l.name: l for l in languages_list}
+        category_dict = {c.name: c for c in categories_list}
+    except Exception as e:
+        print(e)
+
+    # Step 4: Create/Update each resource in the DB
     for resource in unique_resources:
-        languages = resource.get('languages') or []
-        category = resource.get('category')
-
-        for language in languages:
-            if language not in language_dict:
-                language_dict[language] = Language(name=language)
-
-        if category not in category_dict:
-            category_dict[category] = Category(name=category)
-
-        try:
-            new_resource = Resource(
-                name=resource['name'],
-                url=resource['url'],
-                category=category_dict[category],
-                paid=resource.get('paid'),
-                notes=resource.get('notes', ''),
-                upvotes=resource.get('upvotes', 0),
-                downvotes=resource.get('downvotes', 0),
-                times_clicked=resource.get('times_clicked', 0))
-
-            for language in languages:
-                new_resource.languages.append(language_dict[language])
-
-            db.session.add(new_resource)
-            db.session.commit()
-        except exc.SQLAlchemyError as e:
-            db.session.rollback()
-            print('Flask SQLAlchemy Exception:', e)
-            template = "An SQLAlchemy exception of type {0} occurred. Arguments:\n{1!r}"
-            print(resource)
-            break
-        except Exception as e:
-            db.session.rollback()
-            print('exception', e)
-            template = "An exception of type {0} occurred. Arguments:\n{1!r}"
-            print(resource)
-            break
+        category = get_category(resource, category_dict) # Note: modifies the category_dict in place (bad?)
+        langs = get_languages(resource, language_dict) # Note: modifies the language_dict in place (bad?)
+        existing_resource = existing_resources.get(resource['url'])
 
+        if existing_resource:
+            update_resource(resource, existing_resource, langs, category)
+        else:
+            create_resource(resource, langs, category)
 
+def remove_duplicates(data):
+    unique_resources = []
+    resource_dict = {}
+    for resource in data:
+        if not resource_dict.get(resource['url']):
+            resource_dict[resource['url']] = True
+            unique_resources.append(resource)
+    return unique_resources
+
+def get_category(resource, category_dict):
+    category = resource.get('category')
+
+    if category not in category_dict:
```

Comment:
try to use the 
`category_dict.get(category, False)` 
pattern here since it's a constant lookup time instead of linear search.